internal/controller/scheduler/scheduler.go

```diff
@@ -118,3 +125,62 @@ func processVMs(vms []v1.VM) ([]v1.VM, WorkerToResources) {
 
 	return unscheduledVMs, workerToResources
 }
+
+func (scheduler *Scheduler) healthCheckingLoopIteration() error {
+	return scheduler.store.Update(func(txn storepkg.Transaction) error {
+		// Retrieve scheduled VMs
+		vms, err := txn.ListVMs()
+		if err != nil {
+			return err
+		}
+
+		var scheduledVMs []v1.VM
+
+		for _, vm := range vms {
+			if vm.Worker != "" {
+				scheduledVMs = append(scheduledVMs, vm)
+			}
+		}
+
+		// Retrieve and index workers by name
+		workers, err := txn.ListWorkers()
+		if err != nil {
+			return err
+		}
+
+		nameToWorker := map[string]v1.Worker{}
+		for _, worker := range workers {
+			nameToWorker[worker.Name] = worker
+		}
+
+		// Process scheduled VMs
+		for _, scheduledVM := range scheduledVMs {
+			if err := healthCheckVM(txn, nameToWorker, scheduledVM); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+func healthCheckVM(txn storepkg.Transaction, nameToWorker map[string]v1.Worker, vm v1.VM) error {
+	worker, ok := nameToWorker[vm.Worker]
+	if !ok {
+		vm.Status = v1.VMStatusFailed
+		vm.StatusMessage = "VM is assigned to a worker that " +
+			"doesn't exist anymore"
+
+		return txn.SetVM(vm)
+	}
+
+	if worker.Offline() {
```

Comment:
let's check if a vm is not in final state here.

Reply:
See 922e339.